### PR TITLE
Source plugin.

### PR DESCRIPTION
**Pull request by pv101**

### CODE REVIEW

File: pyfr/backends/openmp/provider.py [hunk 1]
```diff
@@ -72,7 +72,6 @@ def _build_function(self, name, src, argtypes, restype=None):
     def _build_kernel(self, name, src, argtypes):
         lib = self._build_library(src)
         fun = lib.function(name, None, [c_void_p])
-
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/provider.py:75]: ?

File: pyfr/backends/base/types.py [hunk 2]
```diff
@@ -122,8 +122,10 @@ def slice(self, ra=None, rb=None, ca=None, cb=None):
 
 
 class Matrix(MatrixBase):
-    def __init__(self, backend, ioshape, initval, extent, aliases, tags):
-        super().__init__(backend, backend.fpdtype, ioshape, initval, extent,
+    def __init__(self, backend, ioshape, initval, extent, aliases, tags,
```
Comment (FreddieWitherden) [inline pyfr/backends/base/types.py:125]: Why is this needed?

File: pyfr/plugins/__init__.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 from pyfr.plugins.pseudostats import PseudoStatsPlugin
 from pyfr.plugins.residual import ResidualPlugin
 from pyfr.plugins.sampler import SamplerPlugin
+from pyfr.plugins.source import Source
```
Comment (FreddieWitherden) [inline pyfr/plugins/__init__.py:11]: SourcePlugin or SourceTermPlugin?

File: pyfr/plugins/source.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        super().__init__(intg, cfgsect, suffix)
+        
+        convars = intg.system.elementscls.convarmap[self.ndims]
+
+        subs = self.cfg.items('constants')
+        subs |= dict(x='ploc[0]', y='ploc[1]', z='ploc[2]')
+        subs |= dict(abs='fabs', pi=math.pi)
+        subs |= {v: f'u[{i}]' for i, v in enumerate(convars)}
+
+        src_exprs = [self.cfg.getexpr('solver-plugin-source', v, '0',
```
Comment (FreddieWitherden) [inline pyfr/plugins/source.py:22]: Linebreak.

File: pyfr/solvers/base/elements.py [hunk 5]
```diff
@@ -217,6 +197,7 @@ def set_backend(self, backend, nscalupts, nonce, linoff):
         nbytes = self.scal_upts[0].nbytes
         ioshape = self.scal_upts[0].ioshape
         aliases = next((m for m in abufs if m.nbytes >= nbytes), None)
+
```
Comment (FreddieWitherden) [inline pyfr/solvers/base/elements.py:200]: ?

File: pyfr/solvers/base/system.py [hunk 6]
```diff
@@ -62,13 +62,18 @@ def __init__(self, backend, rallocs, mesh, initsoln, nregs, cfg):
...
+        self._has_src_macro = False
+
     def commit(self):
         # Prepare the kernels and any associated MPI requests
         self._gen_kernels(self.nregs, self.ele_map.values(), self._int_inters,
                           self._mpi_inters, self._bc_inters)
         self._gen_mpireqs(self._mpi_inters)
         self.backend.commit()
 
+        self._has_src_macro = any([eles._has_src_macro
```
Comment (FreddieWitherden) [inline pyfr/solvers/base/system.py:74]: No need for [].

File: pyfr/solvers/baseadvec/elements.py [hunk 7]
```diff
@@ -108,19 +108,31 @@ def set_backend(self, backend, nscalupts, nonce, linoff):
...
         # Transformed to physical divergence kernel + source term
-        plocupts = self.ploc_at('upts') if plocsrc else None
-        solnupts = self._scal_upts_cpy if solnsrc else None
 
-        if solnsrc:
-            kernels['copy_soln'] = lambda uin: self._be.kernel(
-                'copy', self._scal_upts_cpy, self.scal_upts[uin]
-            )
+        def copy_soln(uin):
+            if have_solnsrc():
```
Comment (FreddieWitherden) [inline pyfr/solvers/baseadvec/elements.py:113]: `if self._soln_in_src_macros`

File: pyfr/solvers/baseadvec/elements.py [hunk 8]
```diff
@@ -63,19 +62,20 @@ def set_backend(self, backend, nscalupts, nonce, linoff):
...
-        plocsrc = self._ploc_in_src_exprs or self._ploc_in_src_macros
-        solnsrc = self._soln_in_src_exprs or self._soln_in_src_macros
 
-        # Source term kernel arguments
-        srctplargs = {
-            'ndims': self.ndims,
-            'nvars': self.nvars,
-            'srcex': self._src_exprs
-        }
+        def have_plocsrc():
```
Comment (FreddieWitherden) [inline pyfr/solvers/baseadvec/elements.py:74]: I think these functions can be inlined

**Comment by FreddieWitherden** [comment]
Please squash into a single commit. Also check how multiple instances work.

File: pyfr/solvers/base/system.py [hunk 9]
```diff
@@ -62,13 +62,18 @@ def __init__(self, backend, rallocs, mesh, initsoln, nregs, cfg):
         self._bc_inters = self._load_bc_inters(rallocs, mesh, elemap)
         backend.commit()
 
+        self._has_src_macro = False
```
Comment (FreddieWitherden) [inline pyfr/solvers/base/system.py:65]: Make this a property.

File: pyfr/plugins/source.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
+import math
+import numpy as np
+import re
+
+from pyfr.plugins.base import BaseSolverPlugin
+
```
Comment (FreddieWitherden) [inline pyfr/plugins/source.py:6]: Add newline.

File: pyfr/backends/base/makoutil.py [hunk 11]
```diff
@@ -84,8 +84,10 @@ def _locals(body):
     # Extract the variable names
     lvars = [re.match(r'\s*(\w+)', v)[1] for v in decls]
 
+    keywords = {'if', 'int'}
```
Comment (FreddieWitherden) [inline pyfr/backends/base/makoutil.py:87]: Is this needed?

File: pyfr/solvers/baseadvec/elements.py [hunk 12]
```diff
@@ -13,24 +14,20 @@ def __init__(self, *kargs, **kwargs):
...
             'ndims': self.ndims,
             'nvars': self.nvars,
-            'srcex': self._src_exprs,
             'srcmacros': []
         }
 
         self._ploc_in_src_macros = False
         self._soln_in_src_macros = False
 
+        self._has_src_macro = False
```
Comment (FreddieWitherden) [inline pyfr/solvers/baseadvec/elements.py:23]: Make this a property `return bool(self._srctplargs['srcmacros'])`

File: pyfr/solvers/baseadvec/kernels/evalsrcmacros.mako [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+% for mod, name in srcmacros:
+    <%include file='${mod}'/>
+% endfor
+
+<%pyfr:kernel name='evalsrcmacros' ndim='2'
+              t='scalar fpdtype_t'
+              ploc='in fpdtype_t[${str(ndims)}]'
+              u='inout fpdtype_t[${str(nvars)}]'
+              >
```
Comment (FreddieWitherden) [inline pyfr/solvers/baseadvec/kernels/evalsrcmacros.mako:11]: Move up.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 14]
```diff
@@ -199,18 +199,26 @@ def restrict(self, l1, l2):
...
+        rtemp = 0 if l1idxcurr == 1 else 1
+
         # Restrict the physical source term
         l1src = self.pintgs[l1]._source_regidx
         l2dst = self.pintgs[l2]._source_regidx
+
+        # If at top level evaluate src macros
+        if l1 == self._order and self.system._has_src_macro:
+            self._add(0, rtemp, 1, l1idxcurr)
+            self.pintgs[l1].system.evalsrcmacros(rtemp)
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:212]: Can we use a shorthand here?

File: pyfr/plugins/source.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        src_exprs = [self.cfg.getexpr('solver-plugin-source', v, '0',
+                     subs=subs) for v in convars]
+
+        ploc_in_src = any(re.search(r'\bploc\b', ex) for ex in src_exprs)
+        soln_in_src = any(re.search(r'\bu\b', ex) for ex in src_exprs)
+
+        if src_exprs:
+            for etype, eles in intg.system.ele_map.items():
+                eles.add_src_macro('pyfr.plugins.kernels.source','source',
```
Comment (FreddieWitherden) [inline pyfr/plugins/source.py:30]: Spacing.

File: pyfr/plugins/kernels/source.mako [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%pyfr:macro name='source' params='t, u, ploc, src'>
+  % for i, ex in enumerate(srcexprs):
```
Comment (FreddieWitherden) [inline pyfr/plugins/kernels/source.mako:4]: Indentation.

File: pyfr/plugins/source.py (hunk 4 above)
Comment (pv101) [reply pyfr/plugins/source.py:22]: Please elaborate, thanks.

File: pyfr/plugins/source.py (hunk 15 above)
Comment (pv101) [reply pyfr/plugins/source.py:30]: Please elaborate, thanks.

File: pyfr/integrators/dual/pseudo/multip.py (hunk 14 above)
Comment (pv101) [reply pyfr/integrators/dual/pseudo/multip.py:212]: Please elaborate, thanks.

File: pyfr/solvers/base/system.py (hunk 9 above)
Comment (pv101) [reply pyfr/solvers/base/system.py:65]: Slightly unsure as to how best to do this. As I wrote it self._has_src_macro has to be set in commit() on L74 below, after kernels are generated but before ele_map is deleted. So do I need to keep the private self._has_src_macro about, and then have a property has_src_macro that returns _has_src_macro? And/or do I need to define a setter method for the property?

File: pyfr/backends/base/makoutil.py (hunk 11 above)
Comment (pv101) [reply pyfr/backends/base/makoutil.py:87]: No, its for turbulence plugin. Will remove.

File: pyfr/backends/base/types.py (hunk 2 above)
Comment (pv101) [reply pyfr/backends/base/types.py:125]: Its for turbulence plugin. Will remove.

File: pyfr/backends/base/backend.py [hunk 17]
```diff
@@ -136,8 +136,10 @@ def const_matrix(self, initval, dtype=None, tags=set()):
 
     @recordmat
     def matrix(self, ioshape, initval=None, extent=None, aliases=None,
-               tags=set()):
-        return self.matrix_cls(self, ioshape, initval, extent, aliases, tags)
+               tags=set(), dtype=None):
```
Comment (pv101) [inline pyfr/backends/base/backend.py:139]: I'll also remove this.

File: pyfr/solvers/base/system.py (hunk 9 above)
Comment (FreddieWitherden) [reply pyfr/solvers/base/system.py:65]: From the variables which we already have we can always determine if we have an src macro or not.  Thus we can make this a property rather than its own variable.

File: pyfr/integrators/dual/pseudo/multip.py (hunk 14 above)
Comment (FreddieWitherden) [reply pyfr/integrators/dual/pseudo/multip.py:212]: What does `self.pintg` evaluate to?

File: doc/src/user_guide.rst [hunk 18]
```diff
@@ -1590,6 +1531,65 @@ Example::
...
+    *string*
+
+9. ``w`` --- w-velocity source term for ``ac-euler`` |
+   ``ac-navier-stokes``:
+
+    *string*
+
+Example::
+
+    [solver-source-terms]
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:1586]: This needs updating.

File: doc/src/user_guide.rst [hunk 19]
```diff
@@ -1590,6 +1531,65 @@ Example::
     int-E = rho*(u*u + v*v + w*w)
     int-enst = rho*(%(vor1)s*%(vor1)s + %(vor2)s*%(vor2)s + %(vor3)s*%(vor3)s)
 
+[solver-plugin-source]
+^^^^^^^^^^^^^^^^^^^^^
+
+Parameterises solution, space (x, y, [z]), and time (t) dependent
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:1537]: Maybe reword?  As the plugin adds in such terms.

File: pyfr/backends/openmp/provider.py [hunk 20]
```diff
@@ -72,7 +72,7 @@ def _build_function(self, name, src, argtypes, restype=None):
...
+
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/provider.py:75]: Empty lines should not be indented.

File: pyfr/plugins/kernels/source.mako [hunk 21]
```diff
@@ -0,0 +1,7 @@
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%pyfr:macro name='source' params='t, u, ploc, src'>
+    % for i, ex in enumerate(srcexprs):
```
Comment (FreddieWitherden) [inline pyfr/plugins/kernels/source.mako:4]: Indentation.  (Usually we do not indent the Mako stuff.)

File: pyfr/plugins/source.py [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+from pyfr.plugins.base import BaseSolverPlugin
+
+
+class SourcePlugin(BaseSolverPlugin):
+    name = 'source'
+    systems = ['*']
+    formulations = ['dual', 'std']
+
+    def __init__(self, intg, cfgsect, suffix = None):
```
Comment (FreddieWitherden) [inline pyfr/plugins/source.py:14]: Remove suffix stuff.

File: pyfr/solvers/base/system.py [hunk 23]
```diff
@@ -61,14 +61,17 @@ def __init__(self, backend, rallocs, mesh, initsoln, nregs, cfg):
         self._mpi_inters = self._load_mpi_inters(rallocs, mesh, elemap)
         self._bc_inters = self._load_bc_inters(rallocs, mesh, elemap)
         backend.commit()
-
+
```
Comment (FreddieWitherden) [inline pyfr/solvers/base/system.py:64]: Indentation.

File: pyfr/solvers/baseadvec/kernels/evalsrcmacros.mako [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+% for mod, name in src_macros:
+    <%include file='${mod}'/>
+% endfor
+
+<%pyfr:kernel name='evalsrcmacros' ndim='2'
+              t='scalar fpdtype_t'
+              ploc='in fpdtype_t[${str(ndims)}]'
+              u='inout fpdtype_t[${str(nvars)}]'>
+
```
Comment (FreddieWitherden) [inline pyfr/solvers/baseadvec/kernels/evalsrcmacros.mako:11]: Remove newline.

File: pyfr/plugins/source.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+        subs |= {v: f'u[{i}]' for i, v in enumerate(convars)}
+
+        src_exprs = [self.cfg.getexpr(cfgsect, v, subs=subs) for v in convars]
+
+        ploc_in_src = any(re.search(r'\bploc\b', ex) for ex in src_exprs)
+        soln_in_src = any(re.search(r'\bu\b', ex) for ex in src_exprs)
+
+        for etype, eles in intg.system.ele_map.items():
+            eles.add_src_macro('pyfr.plugins.kernels.source', 'source', 
+                               {'srcexprs': src_exprs}, ploc=ploc_in_src,
```
Comment (FreddieWitherden) [inline pyfr/plugins/source.py:31]: Do we want to change the key here fro `src_exprs`?

File: pyfr/plugins/source.py [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+
+        ploc_in_src = any(re.search(r'\bploc\b', ex) for ex in src_exprs)
+        soln_in_src = any(re.search(r'\bu\b', ex) for ex in src_exprs)
+
+        for etype, eles in intg.system.ele_map.items():
+            eles.add_src_macro('pyfr.plugins.kernels.source', 'source', 
+                               {'srcexprs': src_exprs}, ploc=ploc_in_src, 
+                               soln=soln_in_src)
+
+    def __call__(self, intg):
```
Comment (FreddieWitherden) [inline pyfr/plugins/source.py:34]: Is this not already defaulted?
